lib/mock/mock-agent.js

```diff
@@ -0,0 +1,185 @@
+const MockClient = require('./mock-client')
+const MockPool = require('./mock-pool')
+const { matchValue, checkNetConnect } = require('./mock-utils')
+const { InvalidArgumentError } = require('../core/errors')
+
+class MockAgent extends Agent {
+  constructor (opts) {
+    super(opts)
+    const mockAgent = this
+
+    this[kOriginalClientDispatch] = Client.prototype.dispatch
+    this[kOriginalPoolDispatch] = Pool.prototype.dispatch
+    this[kNetConnect] = true
+    this[kIsMockActive] = true
+    this[kOriginalClose] = this.close
+    this.close = this[kClose]
+
+    if (this[kIsMockEnabled]()) {
+      Client.prototype.dispatch = this[kBuildMockDispatch](this[kOriginalClientDispatch])
+      Pool.prototype.dispatch = this[kBuildMockDispatch](this[kOriginalPoolDispatch])
+    }
+
+    setGlobalAgent(mockAgent)
+  }
+
+  get (origin) {
+    let pool = this[kMockAgentGet](origin)
+    if (!pool) {
+      pool = this[kCreatePool](origin)
+      this[kMockAgentSet](origin, pool)
+    }
+
+    const originalDispatch = this[kIsClient]() ? this[kOriginalClientDispatch] : this[kOriginalPoolDispatch]
+    const isMockEnabled = this[kIsMockEnabled].bind(this)
+    const getNetConnect = this[kGetNetConnect].bind(this)
+
+    pool.dispatch = function dispatch (opts, handler) {
+      if (isMockEnabled()) {
+        let foundMockDispatch = false
+        foundMockDispatch = pool[kDispatch](opts, handler)
+        if (!foundMockDispatch) {
+          const netConnect = getNetConnect()
+          if (checkNetConnect(netConnect, origin)) {
+            originalDispatch.call(this, opts, handler)
+          } else {
+            throw new Error(`Unable to find mock dispatch and real dispatches are disabled for ${origin}`)
+          }
+        }
+      } else {
+        originalDispatch.call(this, opts, handler)
+      }
+    }
+    return pool
+  }
+
+  deactivate () {
+    this[kIsMockActive] = false
+    Client.prototype.dispatch = this[kOriginalClientDispatch]
+    Pool.prototype.dispatch = this[kOriginalPoolDispatch]
+  }
+
+  activate () {
+    this[kIsMockActive] = true
+    Client.prototype.dispatch = this[kBuildMockDispatch](this[kOriginalClientDispatch])
+    Pool.prototype.dispatch = this[kBuildMockDispatch](this[kOriginalPoolDispatch])
+  }
+
+  enableNetConnect (matcher) {
+    if (typeof matcher === 'string' || typeof matcher === 'function' || matcher instanceof RegExp) {
+      if (Array.isArray(this[kNetConnect])) {
+        this[kNetConnect].push(matcher)
+      } else {
+        this[kNetConnect] = [matcher]
+      }
+    } else if (typeof matcher === 'undefined') {
+      this[kNetConnect] = true
+    } else {
+      throw new InvalidArgumentError('Unsupported matcher. Must be one of String|Function|RegExp.')
+    }
+  }
+
+  disableNetConnect () {
+    this[kNetConnect] = false
+  }
+
+  [kIsMockEnabled] () {
+    if (!this[kIsMockActive]) {
+      return false
+    }
+    return process.env.UNDICI_MOCK_OFF !== 'true'
```

Comment:
Do not use env variables, mocking is always enabled when passing through the mock.

Reply:
Understood!